README.md
![currentcost logo](logo.png)

[![version](https://img.shields.io/github/v/release/lolouk44/CurrentCost_HA_CC)](https://github.com/lolouk44/CurrentCost_HA_CC/releases)
[![hacs_badge](https://img.shields.io/badge/HACS-Custom-orange.svg)](https://github.com/custom-components/hacs)
# CurrentCost Custom Component for Home Assistant

**Please note: Home Assistant 2021.12 or higher is required for this custom component to work**

This repo is a custom component for [Home Assistant](https://www.home-assistant.io/)

The `currentcost` sensor platform is using the data provided by a [CurrentCost](http://www.currentcost.com/) device connected to the serial port via a [data cable](http://www.currentcost.com/product-datacable.html).

The sensor returns the Total Power (usually on Appliance 0) as a state, and the temperature as an attribute.
It is also possible to list additional appliances by listing the appliance number that CurrentCost devices are paired with

Confirmed working devices:
- Currentcost Envi (aka CC128)
- Currentcost EnviR

## HACS Installation

The easiest way to install this custom component is via [HACS](https://hacs.xyz/)
1) Follow the [installation instructions](https://hacs.xyz/docs/installation/prerequisites) to install HACS
2) Click on the HACS icon in the left side bar, click on the elipsis on the right handside and select `Custom Repositories`
3) enter `https://github.com/lolouk44/CurrentCost_HA_CC` in the URL box, select `Integration` as a category


## Manual Installation
To install the CurrentCost custom component:
1) Create a folder called `custom_components` in your config folder (same folder where configuration.yaml is locate, if that folder does not already exist)
2) Create a folder called `currentcost` (no spaces, lowercase)
3) Copy the files inside the [CurrentCost_HA_CC](https://github.com/lolouk44/CurrentCost_HA_CC/tree/master/custom_components/CurrentCost_HA_CC) folder into the `currentcost` folder


## Configuration

To setup a CurrentCost sensor to your installation:
1) Add the following code to your `configuration.yaml` file under the existing `sensor` and `template` headers. (do not copy/paste the `sensor` or `template` headers). 
If `sensor` or `template` do not already exist, add the code block including the `sensor` or `template` header.

```yaml
# Example configuration.yaml entry
sensor:
  - platform: currentcost
    serial_port: /dev/serial/by-id/usb-Prolific_Technology_Inc._USB-Serial_Controller-if00-port0
    name: Current Cost
    unique_id: abaaa250-fd59-46e1-abd8-07545fb2b297
    baudrate: 57600
    devices:
      - 0
      - 2
      - 9
  # Add this sensor if you want to see data in the energy tab
  - platform: integration
    source: sensor.current_cost
    name: Total Energy
    unit_prefix: k
    round: 2

template:
  - sensor:
      - name: "CurrentCost Temperature"
        unit_of_measurement: '°C'
        state: '{{ state_attr("sensor.current_cost", "Temperature") | float -3 }}' # Manual adjustment of -3°C in case the temp sensor is high than real temperature
        device_class: temperature
        state_class: measurement # Add state_class: measurement for long term statistics are required
  - sensor:
      - name: "CurrentCost Power"
        unit_of_measurement: 'W'
        state: '{{ state_attr("sensor.current_cost", "Appliance 0") }}'
        device_class: power
        state_class: measurement
  - sensor:
      - name: "Dehumidifier Power"
        unit_of_measurement: 'W'
        state: '{{ state_attr("sensor.current_cost", "Appliance 2") }}'
        device_class: power
        state_class: measurement
  - sensor:
      - name: "Total Energy Used Last 24h" # Note: this data is published by the Current Cost device every 2h
        unit_of_measurement: 'KWh'
        state: '{{ state_attr("sensor.current_cost", "Appliance 0 Last 24h") }}'
        device_class: energy
```


### serial_port:
**description**: Local serial port where the sensor is connected and access is granted.  
**required**: true  
**type**: string  
note: when using HA in a docker environment, make sure you assign a name to the device when mounting it, e.g. `--device=/dev/ttyUSB0:/dev/ttyUSB0` as opposed to just `--device=/dev/ttyUSB0`
### name:
**description**: Friendly name to use for the frontend. Default to "Current Cost".  
**required**: false  
**type**: string  
### baudrate:
**description**: Baudrate of the serial port. 57600 is the value needed for EnviR devices.  
**required**: false  
**default**: 57600 Bps  
**type**: integer  
### devices:
**description**: List of appliance numbers paired with a CurrentCost sensor  
**required**: false  
**default**: 0  
**type**: integer  


CHANGELOG.md
## [0.2.5] - 2023-06-30
### Changed
- version bump pyserial-asyncio==0.6 to fix for python 3.11 in 2023.6

## [0.2.4] - 2023-01-14
### Added
- Support for unique_id to help sensor customisation
### Changed
- Replaced deprecated DEVICE_CLASS_* Constants ([fixes #20](https://github.com/lolouk44/CurrentCost_HA_CC/issues/20))
- Fixed default icon (flash-circle -> lightning-bolt-circle)

## [0.2.3] - 2022-09-18
### Changed
- Gracefully handle when unable to read serial port ([fixes #17](https://github.com/lolouk44/CurrentCost_HA_CC/issues/17))

## [0.2.2] - 2021-12-14
### Changed
- Updated Sensor definition: replace deprecated device_state_attributes with extra_state_attributes

## [0.2.1] - 2021-08-19
### Added
- Added handling of temperature in Fahrenheit

## [0.2.0] - 2021-08-11
### Added
- Added last 24h and last 30 days energy used in KWh to the attributes

## [0.1.10] - 2021-08-10
### Changed
- Removed "Setting up State Class" error message used for testing

## [0.1.8] - 2021-08-09
### Added
- Updated support for [Long Term Statistics](https://www.home-assistant.io/blog/2021/08/04/release-20218/#long-term-statistics)
### Changed
- Updated README and INFO to use "modern" Template Sensors and allow support for Long Term Statistics for templated sensors

## [0.1.7] - 2021-07-28
### Breaking Change
- Standardized the way sensor data is reported (removed the unit of measurement) - Please update your templates (see README)

## [0.1.6] - 2021-07-27
### Breaking Change
- Standardized the way sensor data is reported (removed the unit of measurement) - Please update your templates (see README)

## [0.1.5] - 2021-03-30
### Changed
- Added version to manifest

## [0.1.4] - 2020-09-10
### Changed
- repo completely overhauled for a PR to make it ready for publishing in HACS

## [0.1.3] - 2020-09-08
### Changed
- Repo structure changed to be HACS Compliant

## [0.1.2] - 2020-09-08
### Changed
- Better error handling: Stop any data processing when error parsing data

## [0.1.1] - 2020-04-03
### Added
- First release (but not tagged on Github)


custom_components/currentcost/sensor.py
"""Support for reading Current Cost data from a serial port."""
import json
import logging
import xmltodict
import sys

import serial_asyncio
import voluptuous as vol

from homeassistant.components.sensor import PLATFORM_SCHEMA, SensorEntity,  SensorDeviceClass, SensorStateClass
from homeassistant.const import CONF_NAME, CONF_UNIQUE_ID, CONF_DEVICES, EVENT_HOMEASSISTANT_STOP, POWER_WATT
import homeassistant.helpers.config_validation as cv

_LOGGER = logging.getLogger(__name__)


CONF_SERIAL_PORT = "serial_port"
CONF_BAUDRATE = "baudrate"
CONF_DEVICES = "devices"

DEFAULT_NAME = "Current Cost"
DEFAULT_ID = "abaaa250-fd59-46e1-abd8-07545fb2b297"
DEFAULT_BAUDRATE = 57600
DEFAULT_DEVICES = [0,1,2,3,4,5,6,7,8,9]

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_SERIAL_PORT): cv.string,
        vol.Optional(CONF_BAUDRATE, default=DEFAULT_BAUDRATE): cv.positive_int,
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
        vol.Optional(CONF_UNIQUE_ID, default=DEFAULT_ID): cv.string,
        vol.Optional(CONF_DEVICES, default=DEFAULT_DEVICES): vol.All(cv.ensure_list, [vol.Range(min=0, max=9)]),
    }
)


async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
    """Set up the Current Cost sensor platform."""
    name = config.get(CONF_NAME)
    unique_id = config.get(CONF_UNIQUE_ID)
    port = config.get(CONF_SERIAL_PORT)
    baudrate = config.get(CONF_BAUDRATE)
    devices = config.get(CONF_DEVICES)
    _LOGGER.debug("devices: %s", config.get(CONF_DEVICES))
    #sensor = []
    sensor = CurrentCostSensor(name, f"current-cost-{unique_id}", port, baudrate, devices)
    #for variable in devices:
    #    sensor.append(CurrentCostSensor(f"{name}_appliance_{variable}", port, baudrate))
    #sensor.append(CurrentCostSensor(f"{name}_temperature", port, baudrate))

    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, sensor.stop_serial_read())
    async_add_entities([sensor], True)


class CurrentCostSensor(SensorEntity):
    """Representation of a Current Cost sensor."""

    def __init__(self, name, unique_id, port, baudrate, devices):
        """Initialize the Current Cost sensor."""
        self._name = name
        self._attr_unique_id = unique_id
        self._unit = POWER_WATT
        self._icon = "mdi:flash-outline"
        self._device_class = SensorDeviceClass.POWER
        self._state_class = SensorStateClass.MEASUREMENT
        self._state = None
        self._port = port
        self._baudrate = baudrate
        self._serial_loop_task = None
        self._attributes = {"Temperature": None}
        self._devices = devices
        for variable in devices:
            self._attributes[f"Appliance {variable}"] = None
            self._attributes[f"Appliance {variable} Last 24h"] = None
            self._attributes[f"Appliance {variable} Last 30 days"] = None

    async def async_added_to_hass(self):
        """Handle when an entity is about to be added to Home Assistant."""
        self._serial_loop_task = self.hass.loop.create_task(
            self.serial_read(self._port, self._baudrate)
        )

    async def serial_read(self, device, rate, **kwargs):
        """Read the data from the port."""
        reader, _ = await serial_asyncio.open_serial_connection(
            url=device, baudrate=rate, **kwargs
        )
        while True:
            try:
                line = await reader.readline()
                line = line.decode("utf-8").strip()
                _LOGGER.debug("Line Received: %s", line)
            except Exception as error:
                _LOGGER.error("Error Reading From Serial Port: %s", error)
                pass
            try:
                data = xmltodict.parse(line)
                # Data can be parsed from line, continuing
                # First read real time data
                try:
                    appliance = int(data['msg']['sensor'])
                except:
                    appliance = None
                    pass
                
                temperature = None
                try:
                    temperature = float(data['msg']['tmpr'])
                except:
                    pass
                try:
                    temperature = float(data['msg']['tmprF'])
                except:
                    pass
                try:
                    imp = int(data['msg']['imp'])
                    ipu = int(data['msg']['ipu'])
                except:
                    imp = None
                    ipu = None
                    pass
                try:
                    wattsch1 = int(data['msg']['ch1']['watts'])
                except:
                    wattsch1 = 0
                    pass
                try:
                    wattsch2 = int(data['msg']['ch2']['watts'])
                except:
                    wattsch2 = 0
                    pass
                try:
                    wattsch3 = int(data['msg']['ch3']['watts'])
                except:
                    wattsch3 = 0
                    pass
                total_watts = wattsch1 + wattsch2 + wattsch3
                if appliance == 0:
                    self._state = total_watts
                    self._attributes[f"Channel 1"] = wattsch1
                    self._attributes[f"Channel 2"] = wattsch2
                    self._attributes[f"Channel 3"] = wattsch3
                if appliance is not None:
                    if imp is not None:
                        self._attributes[f"Impulses {appliance}"] = imp
                        self._attributes[f"Impulses/Unit {appliance}"] = ipu
                    else:
                        self._attributes[f"Appliance {appliance}"] = total_watts
                if temperature is not None:
                    self._attributes["Temperature"] = temperature
                    
                # Then read history data

                for variable in self._devices:
                    #self._attributes[f"Appliance {variable}"] = None
                    try:
                        if int(data['msg']['hist']['data'][int(variable)]['sensor']) == int(variable):
                            applianceHist = int(data['msg']['hist']['data'][int(variable)]['sensor'])
                        else:
                            applianceHist = None
                    except:
                        applianceHist = None
                        pass
                    if applianceHist is not None:
                        try:
                            last24h = float(data['msg']['hist']['data'][applianceHist]['d001'])
                            self._attributes[f"Appliance {applianceHist} Last 24h"] = last24h
                        except:
                            pass
                        try:
                            last30d = float(data['msg']['hist']['data'][applianceHist]['m001'])
                            self._attributes[f"Appliance {applianceHist} Last 30 days"] = last30d
                        except:
                            pass
                    
                # Then update HA Sensor info
                self.async_schedule_update_ha_state()

            # Data can not be parsed from line, raising exception
            except:
                _LOGGER.error(f"Error parsing data from serial port:\n    {sys.exc_info()[1]}\n    line received:\n    {line}")
                pass

    async def stop_serial_read(self):
        """Close resources."""
        if self._serial_loop_task:
            self._serial_loop_task.cancel()

    @property
    def name(self):
        """Return the name of the sensor."""
        return self._name

    @property
    def should_poll(self):
        """No polling needed."""
        return False

    @property
    def extra_state_attributes(self):
        """Return the attributes of the entity (if any JSON present)."""
        return self._attributes

    @property
    def state(self):
        """Return the state of the sensor."""
        return self._state

    @property
    def unit_of_measurement(self):
        """Return the units of measurement."""
        return self._unit

    @property
    def icon(self):
        """Return the icon of the sensor."""
        return self._icon

    @property
    def device_class(self):
        """Return the device class of the sensor."""
        return self._device_class

    @property
    def state_class (self):
        """Return the state class of the sensor."""
        return self._state_class
